File: src/autobot/templates/claude/hooks/session-context.py
#!/usr/bin/env python3
"""
Autobot Session Context Hook - Loads state at session start.

Triggered on SessionStart to provide Claude with current task context,
progress status, and any important learnings from previous iterations.

Hook Output Protocol:
  - Print JSON with "hookSpecificOutput" to inject context
"""
import json
import os
import sys
from pathlib import Path


def load_json(path: Path) -> dict | None:
    """Load JSON file if it exists."""
    if path.exists():
        try:
            return json.loads(path.read_text(encoding='utf-8'))
        except (OSError, json.JSONDecodeError):
            return None
    return None


def get_project_dir() -> Path:
    """Get the project directory from environment or cwd."""
    return Path(os.environ.get('CLAUDE_PROJECT_DIR', os.getcwd()))


def format_progress_bar(completed: int, total: int, width: int = 10) -> str:
    """Create a visual progress bar."""
    if total == 0:
        return "[" + "-" * width + "]"
    filled = int(width * completed / total)
    return "[" + "#" * filled + "-" * (width - filled) + "]"


def main():
    project_dir = get_project_dir()
    autobot_dir = project_dir / '.autobot'

    context_parts = []

    # Check if autobot is set up
    if not autobot_dir.exists():
        print(json.dumps({
            "hookSpecificOutput": {
                "hookEventName": "SessionStart",
                "additionalContext": "Autobot: No active task. Use /init-task to start."
            }
        }))
        sys.exit(0)

    # Load current task
    task = load_json(autobot_dir / 'task.json')
    if task and task.get('id') and task.get('status') != 'idle':
        context_parts.append(f"ACTIVE TASK: {task.get('title', 'Unknown')}")
        context_parts.append(f"STATUS: {task.get('status', 'unknown')}")
    else:
        print(json.dumps({
            "hookSpecificOutput": {
                "hookEventName": "SessionStart",
                "additionalContext": "Autobot: No active task. Use /init-task to start."
            }
        }))
        sys.exit(0)

    # Load plan status
    plan = load_json(autobot_dir / 'plan.json')
    if plan and plan.get('subtasks'):
        subtasks = plan.get('subtasks', [])
        completed = len([t for t in subtasks if t.get('status') == 'completed'])
        in_progress = [t for t in subtasks if t.get('status') == 'in_progress']
        pending = [t for t in subtasks if t.get('status') == 'pending']
        total = len(subtasks)

        progress_bar = format_progress_bar(completed, total)
        context_parts.append(f"PROGRESS: {progress_bar} {completed}/{total} subtasks")

        # Show current/next subtask
        if in_progress:
            context_parts.append(f"CURRENT: {in_progress[0].get('title')}")
        elif pending:
            context_parts.append(f"NEXT: {pending[0].get('title')}")

    # Load metrics
    metrics = load_json(autobot_dir / 'metrics.json')
    if metrics:
        iteration = metrics.get('current_iteration', 0)
        max_iter = metrics.get('max_iterations', 50)
        context_parts.append(f"ITERATION: {iteration}/{max_iter}")

        # Warn about consecutive failures
        failures = metrics.get('consecutive_test_failures', 0)
        if failures > 0:
            context_parts.append(f"WARNING: {failures} consecutive test failures")

    # Check for pause flag
    if (autobot_dir / '.paused').exists():
        context_parts.append("STATUS: PAUSED (use /resume to continue)")

    # Load recent learnings from progress.md (last 300 chars)
    progress_path = autobot_dir / 'progress.md'
    if progress_path.exists():
        progress = progress_path.read_text(encoding='utf-8')
        # Find recent entries (after the header)
        if '---' in progress:
            recent = progress.split('---')[-1].strip()
            if recent and len(recent) > 10:
                # Truncate if too long
                if len(recent) > 300:
                    recent = "..." + recent[-300:]
                context_parts.append(f"\nRECENT LEARNINGS:\n{recent}")

    # Output context
    if context_parts:
        print(json.dumps({
            "hookSpecificOutput": {
                "hookEventName": "SessionStart",
                "additionalContext": "\n".join(context_parts)
            }
        }))


if __name__ == "__main__":
    main()

File: .claude/hooks/session-context.js
#!/usr/bin/env node
/**
 * Autobot Session Context Hook - Loads state at session start.
 *
 * Triggered on SessionStart to provide Claude with current task context,
 * progress status, and any important learnings from previous iterations.
 */
import fs from 'fs';
import path from 'path';

function loadJson(filePath) {
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
  } catch {
    return null;
  }
}

function getProjectDir() {
  return process.env.CLAUDE_PROJECT_DIR || process.cwd();
}

function formatProgressBar(completed, total, width = 10) {
  if (total === 0) {
    return '[' + '-'.repeat(width) + ']';
  }
  const filled = Math.floor(width * completed / total);
  return '[' + '#'.repeat(filled) + '-'.repeat(width - filled) + ']';
}

function main() {
  const projectDir = getProjectDir();
  const autobotDir = path.join(projectDir, '.autobot');
  const contextParts = [];

  if (!fs.existsSync(autobotDir)) {
    console.log(JSON.stringify({
      hookSpecificOutput: {
        hookEventName: 'SessionStart',
        additionalContext: 'Autobot: No active task. Use /init-task to start.',
      },
    }));
    process.exit(0);
  }

  // Load current task
  const task = loadJson(path.join(autobotDir, 'task.json'));
  if (task && task.id && task.status !== 'idle') {
    contextParts.push(`ACTIVE TASK: ${task.title || 'Unknown'}`);
    contextParts.push(`STATUS: ${task.status || 'unknown'}`);
  } else {
    console.log(JSON.stringify({
      hookSpecificOutput: {
        hookEventName: 'SessionStart',
        additionalContext: 'Autobot: No active task. Use /init-task to start.',
      },
    }));
    process.exit(0);
  }

  // Load plan status
  const plan = loadJson(path.join(autobotDir, 'plan.json'));
  if (plan && plan.subtasks && plan.subtasks.length > 0) {
    const subtasks = plan.subtasks;
    const completed = subtasks.filter(t => t.status === 'completed').length;
    const inProgress = subtasks.filter(t => t.status === 'in_progress');
    const pending = subtasks.filter(t => t.status === 'pending');
    const total = subtasks.length;

    const progressBar = formatProgressBar(completed, total);
    contextParts.push(`PROGRESS: ${progressBar} ${completed}/${total} subtasks`);

    if (inProgress.length > 0) {
      contextParts.push(`CURRENT: ${inProgress[0].title}`);
    } else if (pending.length > 0) {
      contextParts.push(`NEXT: ${pending[0].title}`);
    }
  }

  // Load metrics
  const metrics = loadJson(path.join(autobotDir, 'metrics.json'));
  if (metrics) {
    const iteration = metrics.current_iteration || 0;
    const maxIter = metrics.max_iterations || 50;
    contextParts.push(`ITERATION: ${iteration}/${maxIter}`);

    const failures = metrics.consecutive_test_failures || 0;
    if (failures > 0) {
      contextParts.push(`WARNING: ${failures} consecutive test failures`);
    }
  }

  // Check for pause flag
  if (fs.existsSync(path.join(autobotDir, '.paused'))) {
    contextParts.push('STATUS: PAUSED (use /resume to continue)');
  }

  // Load recent learnings from progress.md
  const progressPath = path.join(autobotDir, 'progress.md');
  if (fs.existsSync(progressPath)) {
    const progress = fs.readFileSync(progressPath, 'utf-8');
    if (progress.includes('---')) {
      let recent = progress.split('---').pop().trim();
      if (recent && recent.length > 10) {
        if (recent.length > 300) {
          recent = '...' + recent.slice(-300);
        }
        contextParts.push(`\nRECENT LEARNINGS:\n${recent}`);
      }
    }
  }

  if (contextParts.length > 0) {
    console.log(JSON.stringify({
      hookSpecificOutput: {
        hookEventName: 'SessionStart',
        additionalContext: contextParts.join('\n'),
      },
    }));
  }
}

main();
